### PR TITLE
[Montecarlo][11580] Make TDatabasePDG::GetParticle methods thread safe

### LINKED ISSUE #11580: TDatabasePDG::GetParticle(...) not being thread safe
### Describe the bug
concurrent call to methods like `TDatabasePDG::Instance()->GetParticle(xxx)->GetName()` can cause segmentation violation. Even with `ROOT::EnableImplicitMT()` Set

### Expected behavior
ROOT::EnableImplicitMT() should make TDatabasePDG::GetParticle(...) thread safe without user taking locks

### To Reproduce
Any program calling `TDatabasePDG::Instance()->GetParticle(xxx)->GetName()` at different thread

### Setup
```
ROOT Version: 6.26/06
Platform: Fedora 37 Beta
Compiler: g++ (GCC) 12.2.1 20220819 (Red Hat 12.2.1-2)
```

### Additional context
https://root-forum.cern.ch/t/tdatabasepdg-is-not-thread-safe-while-creating-multiple-thread-only-instance-cause-log-spam/51957

### PR DESCRIPTION
**Pull request by dpiparo**
# This Pull request:

## Changes or fixes:
Makes TDatabasePDG::GetParticle methods thread safe and adds a test for this feature.

## Checklist:

- [v] tested changes locally
- [ ] updated the docs (if necessary)

This PR fixes #11580

### CODE REVIEW

File: montecarlo/eg/test/pdg.cxx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      "g", "gamma", "Z0", "W+", "h0", "reggeon", "pomeron", "Z'0", "Z\"0", "W'+", "H0", "A0", "H+",
+      // Leptons
+      "e-", "nu_e", "mu-", "nu_mu", "tau-", "nu_tau", "tau'-", "nu'_tau"};
+
+   std::vector<int> fParticleCodes;
+   TDatabasePDG *fInstance;
+   ROOT::TThreadExecutor fPool;
+};
+
+TEST_F(TPDGTable, ConcurrentGetParticleByNameCalls)
```
Comment (pcanal) [inline montecarlo/eg/test/pdg.cxx:59]: I am not clear on the semantic of `TEST_F` here but wouldn't this means that the `TPDGTable` object will be created first and isn't the constructor of `TPDGTable` already (indirectly) initializing the pdg table?  I.e. Do we test what we mean to test here?

Comment (dpiparo) [reply montecarlo/eg/test/pdg.cxx:59]: hi @pcanal. You are absolutely right. I think I understood from the report that the 2 GetParticle methods needed to be thread safe and this is what this PR enables and tests. Making the static function `TDatabasePDG::Instance()` also ts is of course possible, even if I think was not requested (and somewhat feels like swimming against the current, given that having TS getters is perhaps natural, while initialisation less so, in general). Of course I am open to any suggestion, given that technically the cost to implement more features is quite low.

Comment (pcanal) [reply montecarlo/eg/test/pdg.cxx:59]: >? Making the static function TDatabasePDG::Instance() also ts is of course possible

It of course should be but indeed should not be really part of this PR :).

What I observe is that the test is currently not testing the thread safety of `GetParticle`.

`TPDGTable` is created first (before threads) and its constructor contains:
```
this->fInstance->GetParticle(pname)->PdgCode();
```
consequently, (If I read correctly), `TDatabasePDG::fParticleList` will be set then and subsequent calls to `GetParticle` will work properly from multiple threads with or without this PR.

Am I missing something?

**Comment by dpiparo** [comment]
This version of the PR is the result of some thinking and two test implementations, one with locks and one relying on TThreadedObject. Both solutions seem heavy and have impacts on runtime. The proposed changes, on the other hand, preserve the existing behaviour and allow thread safe access to the db, documenting well the properties of the TDatabasePDG component.

File: montecarlo/eg/test/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# Copyright (C) 1995-2023, Rene Brun and Fons Rademakers.
+# All rights reserved.
+#
+# For the licensing terms see $ROOTSYS/LICENSE.
+# For the list of contributors see $ROOTSYS/README/CREDITS.
+
+# @author Danilo Piparo CERN
+
+#### C++ TESTS ####
+ROOT_ADD_GTEST(pdgdatabase pdg.cxx LIBRARIES EG)
```
Comment (Axel-Naumann) [inline montecarlo/eg/test/CMakeLists.txt:10]: Could you please add a training newline?

File: montecarlo/eg/src/TDatabasePDG.cxx [hunk 3]
```diff
@@ -100,6 +100,33 @@ TDatabasePDG::~TDatabasePDG()
...
+///   auto getParticle = [](const char *pname) {
+///   auto particle = TDatabasePDG::Instance()->GetParticle(pname);
+///   // Do something with particle
+///   };
+///   ROOT::TThreadExecutor pool(4);
+///   mypool.Foreach(getParticle, {"g", "gamma", "Z0", "W+", "h0"}});
+///~~~
+
+void TDatabasePDG::Init()
+{
```
Comment (Axel-Naumann) [inline montecarlo/eg/src/TDatabasePDG.cxx:122]: Would it make sense to move this initialization into the static `Instance` method which guarantees that the initialization happens exactly once, during the initialization of the static local instance variable? Then any other method such as `TDatabasePDG::GetParticle()` would automatically be thread safe, without any changes to user code.

File: montecarlo/eg/test/pdg.cxx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+#ifdef R__USE_IMT
+#include <ROOT/TThreadExecutor.hxx>
+#endif
+#include <TROOT.h>
+
+#include "gtest/gtest.h"
+
+#include <vector>
+
+#ifndef R__B64
```
Comment (Axel-Naumann) [inline montecarlo/eg/test/pdg.cxx:20]: Can you add a comment why 32bit means single-threaded?

File: montecarlo/eg/test/pdg.cxx [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+TEST(TDatabasePDGMT, Initialization)
+{
+   RunOnThreads(TDatabasePDG::Instance);
+   delete TDatabasePDG::Instance();
+}
+
+TEST(TDatabasePDGMT, GetParticleByName)
+{
+   RunOnThreads([]() { TDatabasePDG::Instance()->GetParticle("pi0"); });
```
Comment (Axel-Naumann) [inline montecarlo/eg/test/pdg.cxx:43]: Maybe it's worth checking (`EXPECT_...`) for some property of the pi0 (such as its mass) to have a better guarantee that there is no race in initialization etc? Same for the other tests below.

**Review by Axel-Naumann — APPROVED** [review]
Very nice improvement! But see comment.

**Comment by dpiparo** [comment]
made clang-format happy.

**Review by bellenot — APPROVED** [review]

**Pull request merged**